src/sync_key_gen.rs

```diff
@@ -10,7 +10,89 @@
 //! signatures and encryption. The secret master key is not known by anyone. The protocol succeeds
 //! if up to `threshold` nodes are faulty.
 //!
-//! # How it works
+//! ## Usage example
+//!
+//! ```
+//! extern crate rand;
+//! extern crate hbbft;
+//!
+//! use std::collections::BTreeMap;
+//!
+//! use hbbft::crypto::{PublicKey, SecretKey, SignatureShare};
+//! use hbbft::sync_key_gen::{ProposeOutcome, SyncKeyGen};
+//!
+//! // Two out of four shares will suffice to sign or encrypt something.
+//! let (threshold, node_num) = (1, 4);
+//!
+//! // Generate individual key pairs for encryption. These are not suitable for threshold schemes.
+//! let sec_keys: Vec<SecretKey> = (0..node_num).map(|_| rand::random()).collect();
+//! let pub_keys: BTreeMap<usize, PublicKey> = sec_keys
+//!     .iter()
+//!     .map(SecretKey::public_key)
+//!     .enumerate()
+//!     .collect();
+//!
+//! // Create the `SyncKeyGen` instances. The constructor also outputs the proposal that needs to
+//! // be sent to all other participants, so we save the proposals together with their sender ID.
+//! let mut nodes = BTreeMap::new();
+//! let mut proposals = Vec::new();
+//! for (id, sk) in sec_keys.into_iter().enumerate() {
+//!     let (sync_key_gen, opt_proposal) = SyncKeyGen::new(&id, sk, pub_keys.clone(), threshold);
+//!     nodes.insert(id, sync_key_gen);
+//!     proposals.push((id, opt_proposal.unwrap())); // Would be `None` for observer nodes.
+//! }
+//!
+//! // All nodes now handle the proposals and send the resulting `Accept` messages.
+//! let mut accepts = Vec::new();
+//! for (sender_id, proposal) in proposals {
+//!     for (&id, node) in &mut nodes {
+//!         match node.handle_propose(&sender_id, proposal.clone()) {
+//!             Some(ProposeOutcome::Valid(accept)) => accepts.push((id, accept)),
+//!             Some(ProposeOutcome::Invalid(faults)) => panic!("Invalid proposal: {:?}", faults),
+//!             None => panic!("We are not an observer, so we should send Accept."),
```

Comment:
The wording here is a tiny bit confusing (or at least it was to me at first glance). More detailed documentation on `::handle_propose` (and other methods) explaining why/when `None` is returned would probably be helpful.